### ISSUE TITLE
Couple lr schedule and stopping rule into fixed_budget / open_ended modes

### ISSUE THREAD

**Issue by matthiaskloft** [matthiaskloft/bayesflow-hpo#67]
## Summary

`optimize()` exposes the learning-rate schedule and the stopping rule as independent
settings, and the default combination of the two is not coherent:

- `epochs=200` (default)
- `early_stopping_patience=5` (default)
- `decay_steps = num_batches * epochs` — cosine annealing to zero over the full 200
  epochs (`optimization/objective.py:667`)

A trial that early-stops around epoch 20 has annealed roughly 1% of that horizon, so it
stops at essentially peak learning rate. The returned model is not "the 20-epoch model";
it is a model frozen mid-flight, worse than a run actually scheduled for 20 epochs would
have been. Because early stopping *keeps* the model it stops — unlike pruning, which
discards it — this degrades the trial result rather than merely wasting compute.

Proposal: replace the two independent knobs with one **mode**, so schedule and stopping
are always chosen as a coherent pair, and fix three correctness issues that fall out.

## Proposed modes

|              | `fixed_budget`                            | `open_ended`                           |
| ------------ | ----------------------------------------- | -------------------------------------- |
| lr schedule  | cosine to zero                            | inverse-sqrt with warmup               |
| horizon      | the trial's actual `epochs * num_batches` | none                                   |
| stopping     | runs to budget                            | early stop on the validation objective |
| `epochs`     | derived from a simulation budget          | a generous cap, not a target           |
| pruning      | allowed (discards, so safe)               | allowed                                |

**Invariant to enforce:** annealing to a horizon and stopping before that horizon are
mutually exclusive. Setting `early_stopping_patience` together with a horizon-annealed
schedule should warn or raise.

**Suggested default: `fixed_budget`.** Under early stopping trials receive different
budgets, so "best trial" conflates configuration quality with how much training that
trial happened to get. Equal-budget trials are directly comparable; early-stopped ones
are not. `open_ended` is the right choice when the output is "the best model I can get"
and per-configuration convergence differences are large or unknown; `fixed_budget` is
right when the output is a *ranking* that has to be defended.

## Composing the two modes

The modes are most useful in sequence rather than as an either/or, and the API should
make that cheap (for example a mode switch on an otherwise resumable study):

1. **Screen — `open_ended`.** Wide space, one seed per trial, early stopping. Cheap per
   trial and tolerant of configurations whose convergence speed differs a lot. Its
   output is a *shortlist*, not a ranking — unequal budgets make the ordering unsafe to
   report.
2. **Confirm — `fixed_budget`.** Re-run the top ~5 at equal budget with several seeds,
   under the schedule the final model will actually use. Rank here.

This also addresses a separate problem: within-configuration seed noise in our study is
sd ~0.005-0.009 while the spread across the best architectures is ~0.010, so a
single-seed leaderboard is substantially seed lottery. Putting the replication only in
stage 2 costs roughly `N + 15` runs instead of `3N`, and puts it where the decision is.

It also removes the main objection to `open_ended` ranking under a different schedule
than the final fit: stage 2 re-fits under the production schedule anyway.

### 1. Schedule horizon can diverge from actual training length

```python
# optimization/objective.py:667
decay_steps = config.num_batches * config.epochs   # study-level constants
```

but a custom `train_fn` trains for the per-trial values. In `bayesflow-irt`:

```python
# bayesflow_irt/hpo.py:536-538
epochs=int(hparams["epochs"]),
num_batches=int(hparams["num_batches"]),
```

While these agree the mismatch is invisible. As soon as training length varies per trial
— which `fixed_budget` requires — every trial trains under a schedule annealed for the
wrong length. Fix: derive `decay_steps` from the same values `train_fn` uses.

### 2. Early stopping monitors training loss

`EarlyStoppingOnMA(monitor="loss")` stops on a moving average of the **training** loss.
Measured counter-example from the `bayesflow-irt` 2PL study (30x500 corner, recovery
correlation for `log(a)`):

| configuration               | final training loss | recovery          |
| --------------------------- | ------------------- | ----------------- |
| `v2_irc_sabs2_lr1e3`        | 0.5705              | 0.930             |
| `v2_sabs0_h64b4_bs32_dur2x` | **0.5622** (better) | **0.912** (worse) |

Training loss inverts against the objective across architectures — the axis HPO varies.
Pruning is unaffected: `PeriodicValidationCallback` already reports the validation
objective via `trial.report()`. Early stopping does not, and it keeps the model it
stops. Fix: default early stopping to that same validation objective.

### 3. No derived hyperparameters

In online SBI `batch_size` is simultaneously the data-volume knob:
`simulations = batch_size * steps`. At most two of {batch, steps, simulations} can be
fixed. The meaningful choice is to fix the **simulation budget** — it is proportional to
GPU-seconds, which is what a run costs — and let steps follow:

```
num_batches = sims_budget // (batch_size * epochs)
```

Today that arithmetic can only live in a custom `train_fn`, where it collides with issue
1, since the scheduler would not know about it. A first-class derived dimension would let
the budget be declared and have steps and epochs fall out.

Motivation: tuning `epochs` and `lr` as independent dimensions is close to redundant.
Under cosine annealing the optimizer's total displacement scales roughly with
`lr_peak * steps`, so the two form a ridge — and TPE models parameters marginally, which
is the worst case for a ridge. Measured: `lr=2e-3` at 16 epochs matched `lr=5e-4` at 32
epochs, and 64 epochs added nothing further. Making `epochs` derived removes the ridge
rather than asking the sampler to navigate it.

## Explicitly not requested

Conditional or nested search spaces (an `lr` range depending on the sampled
`batch_size`). The coupling has a known functional form — Smith et al. (2018) show the
optimum scales as `B` proportional to `epsilon` — so the right treatment is
reparametrisation, `lr = lr_ref * (batch_size / reference_batch)`, keeping the space
rectangular. A rectangle in decoupled coordinates is easier for TPE than a conditional
space, not merely easier to implement.

## References

- Smith, S. L., Kindermans, P.-J., Ying, C., & Le, Q. V. (2018). *Don't decay the
  learning rate, increase the batch size.* ICLR.
  https://doi.org/10.48550/arXiv.1711.00489
- Shallue, C. J., Lee, J., Antognini, J. M., Sohl-Dickstein, J., Frostig, R., & Dahl,
  G. E. (2019). Measuring the effects of data parallelism on neural network training.
  *JMLR, 20*(112), 1-49. https://doi.org/10.48550/arXiv.1811.03600 — Sec. 2: a learning
  rate tuned at one batch size and reused at a larger one produces either divergence or
  failure to reach the error goal within a fixed epoch budget.
- Draxler, F., Sorrenson, P., Zimmermann, L., Rousselot, A., & Köthe, U. (2023).
  *Free-form flows.* App. C.1, Tables 3-4 — grid-searches batch size (2^2 to 2^8) jointly
  with learning rate, per dataset *and per simulation budget*, on the SBI benchmark.
  Selected batch grows with budget (median 8 / 32 / 64 across the three budgets), which
  is why batch size should be tunable rather than a constant.

## Status

The 3x3 batch x learning-rate grid this issue was filed ahead of has **completed**
(9 arms x 3 seeds, simulation-matched at 102,400 simulations). Results are in
https://github.com/matthiaskloft/bayesflow-hpo/issues/67#issuecomment-5349211054.

Headline changes to the proposal above, from that data:

- The objective **cannot be recovery correlation alone** — the worst-calibrated cell in the
  grid ranks 4th of 9 on recovery. This reinforces correctness issue 2.
- `batch_size` should not stay `IntDimension(constant=256)`: the batch x lr interaction is
  invisible at constant batch, and it is the largest effect measured.
- The `initial_lr` upper bound of 5e-3 is likely too low, though the higher rate trades
  `b` calibration against `a` calibration at few items rather than dominating.

— Claude Code (Opus 5)

**Comment by matthiaskloft** [comment]
## Grid results

The 3x3 batch x learning-rate grid referenced in the Status section has finished:
9 arms x 3 seeds = 27 runs, all successful, simulation-matched at 102,400 simulations
on a single architecture (`v2_irc_sabs0_h64b4`, 2PL). Learning rate is the peak of the
existing cosine-to-zero schedule with warmup. Metrics are means over 3 seeds.

Seed dominates the level of every metric — seed 2 > seed 3 > seed 1 in all nine arms —
so the comparisons below are read within-seed, not off absolute values.

### Recovery (correlation, `log(a)`, corner I30_P500)

| batch | lr 5e-4 | lr 2e-3 | lr 8e-3 | s/run |
| ----- | ------- | ------- | ------- | ----- |
| 32    | 0.926   | 0.929   | 0.929   | 490   |
| 64    | 0.923   | 0.927   | 0.929   | 338   |
| 128   | 0.915   | 0.924   | 0.926   | 268   |

### Calibration error, `b`

| batch | corner | lr 5e-4 | lr 2e-3 | lr 8e-3 |
| ----- | -------- | ------- | ------- | ------- |
| 32    | I30_P500 | 0.0602  | 0.0156  | 0.0111  |
| 64    | I30_P500 | 0.0765  | 0.0354  | 0.0186  |
| 128   | I30_P500 | 0.1063  | 0.0526  | 0.0298  |
| 32    | I30_P10  | 0.0801  | 0.0197  | 0.0154  |
| 64    | I30_P10  | 0.0998  | 0.0648  | 0.0228  |
| 128   | I30_P10  | 0.1241  | 0.0878  | 0.0456  |

### Calibration error, `log(a)`

| batch | corner  | lr 5e-4 | lr 2e-3 | lr 8e-3 |
| ----- | ------- | ------- | ------- | ------- |
| 32    | I30_P500 | 0.0191 | 0.0101  | 0.0126  |
| 64    | I30_P500 | 0.0257 | 0.0125  | 0.0141  |
| 128   | I30_P500 | 0.0690 | 0.0208  | 0.0145  |
| 32    | I5_P500  | 0.0189 | 0.0188  | 0.0287  |
| 64    | I5_P500  | 0.0212 | 0.0191  | 0.0306  |
| 128   | I5_P500  | 0.0336 | 0.0297  | 0.0391  |

## What this changes in the proposal

### 1. The objective cannot be recovery correlation alone

`bs128 lr5e-4` ranks 4th of 9 on recovery (0.915, within ~0.01 of the best) and **last of 9
on calibration in all eight measured cells** — 0.1063 on `b` at I30_P500 against 0.0298 for
the same batch size at lr 8e-3, and 0.0690 on `log(a)` against ~0.015 elsewhere.

Recovery correlation is simply the insensitive instrument here: a configuration it calls
near-optimal is 3-4x worse calibrated. Any single-objective HPO run on recovery would select
this cell. This strengthens the case in correctness issue 2 for making the *validation
objective* explicit and multi-component, and it is an argument against defaulting the
objective to a point-estimate recovery metric.

### 2. Batch size and learning rate interact as Smith et al. predict, and it shows up in calibration

The penalty for reusing a learning rate tuned at a smaller batch grows with batch size.
On recovery the effect is small but monotone (+0.003 / +0.006 / +0.011 going from lr 5e-4 to
the best lr at batch 32 / 64 / 128). On `b` calibration at I30_P500 the same comparison is
5.4x / 4.1x / 3.6x.

This is direct support for the reparametrisation in "Explicitly not requested":
`lr = lr_ref * (batch_size / reference_batch)` rather than a conditional space. It is also
support for making `batch_size` tunable at all — it is currently `IntDimension(constant=256)`
in `search_spaces/training.py`, and at constant batch this entire interaction is invisible.

### 3. The `initial_lr` upper bound of 5e-3 is probably too low — but 8e-3 is not a free win

lr 8e-3 did not diverge at any batch size, and it is best or near-best on `b` calibration in
every row. But on `log(a)` at the 5-item corner it is the **worst** of the three learning
rates at every batch size (0.0287 / 0.0306 / 0.0391 vs ~0.019 at lr 5e-4).

So the recommendation is to widen the range so the sampler can reach ~8e-3, not to re-centre
on it. The trade-off between `b` calibration and `a` calibration at few items is real, and it
is exactly the kind of thing a fixed cap hides.

### 4. Throughput argues for large batches once lr is rescaled

At its own best learning rate, batch 128 costs 268 s/run against 490 s/run for batch 32 — a
1.83x speedup — for a recovery difference of 0.003, well inside the seed noise (sd ~0.005-0.009).
Under a fixed simulation budget, batch size is therefore mostly a throughput knob rather than a
quality knob, provided the learning rate moves with it. This is the derived-budget argument in
correctness issue 3: fixing the simulation budget and letting steps follow is the parametrisation
that makes this trade-off expressible.

## Caveats

- One architecture, one model (2PL), one schedule (cosine-to-zero with warmup). The interaction
  direction is predicted by theory and reproduced here; the magnitudes are not portable.
- 3 seeds per cell. Seed variance is comparable to the between-cell differences on recovery,
  which is why the calibration table carries the weight of the conclusions.
- This grid does **not** explain an earlier, much larger batch-128 failure (recovery 0.667, 3/8
  runs bad) seen on a *different* architecture. That arm's collapse is not reproduced at
  unrescaled lr here, so it has a separate cause still to be identified. What this grid
  establishes is the direction and the calibration cost, not that explanation.

— Claude Code (Opus 5)

**Comment by matthiaskloft** [comment]
## Addendum: warmup length is an explicitly rejected dimension

Recording this now because the argument is easy to reconstruct backwards and get wrong.

Once the `initial_lr` range widens (see the grid results above — three of the four
Pareto-optimal cells used `8e-3`), warmup becomes tempting to tune. It is currently
hardcoded at 5% of total steps in both `bayesflow-irt`'s notebook and its study runner,
and it has never been swept. Warmup is also the mechanism that makes high learning rates
survivable, so holding it fixed while widening the lr range is exactly the kind of
untested coupling that produced the batch-128 artifact this grid was run to disentangle.

**The argument for tuning it** is Shallue et al. (2019), Sec. 5: "the optimal values of
other metaparameters, such as the momentum and learning rate decay schedule, should not
be assumed to remain constant or scale in a simple way as the batch size increases."

**The argument against is stronger, and comes from the same paper.** In their limitations
(Sec. 5.1):

> "We are most confident in our search spaces that tuned the fewest metaparameters (such
> as in our experiments that only tuned learning rate and momentum). We found it quite
> difficult to be confident that our tuning was sufficient when we searched over learning
> rate decay schedules; readers should be aware that the steps to result measurement is
> generally quite sensitive to the learning rate schedule. Thus, we may not have sampled
> enough trials at some batch sizes or, nearly equivalently, our search spaces may have
> been too wide at some batch sizes."

That is the authors of the definitive batch-size study reporting that adding schedule
parameters is what made their own tuning unreliable — at a budget far larger than ours.
Adding a third correlated axis to `{batch, lr}`, on top of the lr x epochs ridge already
described in correctness issue 3, is the change most likely to degrade the rest of the
search.

**Proposed treatment**: sweep warmup factorially *outside* the HPO and fix it at a
known-good constant, leaving the search space as `{batch, lr}` — the configuration
Shallue expresses most confidence in. This fits the two-stage structure already proposed
here: screen cheaply on a narrow space, confirm on the production schedule. Warmup
belongs to the design of the schedule, not to the per-trial search.

A cheap design that also settles a pending question: warmup fraction
`{0, 0.02, 0.05, 0.15}` crossed with lr `{2e-3, 8e-3}` at one batch size, on the
architecture the notebook actually uses (`num_outer_sabs=2`) — which is the architecture
`8e-3` has not yet been tested on. 8 cells x 3 seeds, roughly 2 GPU-hours.

Tracked in `docs/TODO.md` as Package K, item 5.

— Claude Code (Opus 5)
